src/librustc/mir/mod.rs

```diff
@@ -2464,9 +2464,9 @@ impl<'tcx> TypeFoldable<'tcx> for UserTypeProjection {
             .iter()
             .map(|elem| match elem {
                 Deref => Deref,
-                Field(f, ()) => Field(f.clone(), ()),
+                Field(f, ()) => Field(*f, ()),
                 Index(()) => Index(()),
-                elem => elem.clone(),
+                elem => *elem,
```

Comment:
This entire `.map(|elem| ...)` feels like `.copied()`.
So really, `projs` is just `self.projs.clone()`.

Or better, `self.projs.fold_with(folder)` (and derive/impl `TypeFoldable` on `ProjectionElem` if it doesn't already have that - it should, though).